Incorrect handling variadic parameters

With deepscan and variadic functions, linter exits with panic:
panic: runtime error: index out of range [1] with length 1

goroutine 340 [running]:
github.com/fe3dback/go-arch-lint/internal/services/checker/deepscan.(*Searcher).applyMethodImplementationsInPackages.func1(0x140000f7b00)
        /Users/ilyahontarau/go/src/github.com/ilya-hontarau/go-arch-lint/internal/services/checker/deepscan/searcher_find_implementations.go:137 +0x41c
github.com/fe3dback/go-arch-lint/internal/services/checker/deepscan.(*Searcher).findFunctionCalls.func1({0x10237eb78?, 0x140000f7b00})
        /Users/ilyahontarau/go/src/github.com/ilya-hontarau/go-arch-lint/internal/services/checker/deepscan/searcher_find_implementations.go:245 +0xec
go/ast.inspector.Visit(0x140006406c0, {0x10237eb78?, 0x140000f7b00?})
        /opt/homebrew/opt/go/libexec/src/go/ast/walk.go:386 +0x38
go/ast.Walk({0x102379718?, 0x140006406c0?}, {0x10237eb78, 0x140000f7b00})

--- a/internal/services/checker/deepscan/models.go
+++ b/internal/services/checker/deepscan/models.go
@@ -19,6 +19,7 @@ type (
 		ArgumentDefinition Source           // where method param type defined (func (a,b,c _int_))
 		Interface          Interface        // used interface for injection
 		Implementations    []Implementation // all code links to this param
+		IsVariadic         bool             // function param is variadic
 	}
 
 	Interface struct {

--- a/internal/services/checker/deepscan/models.go
+++ b/internal/services/checker/deepscan/models.go
@@ -19,7 +19,7 @@ type (
 		ArgumentDefinition Source           // where method param type defined (func (a,b,c _int_))
 		Interface          Interface        // used interface for injection
 		Implementations    []Implementation // all code links to this param
-		IsVariadic         bool             // function param is variadic
+		IsVariadic         bool             // function param is variadic (func (a bool, nums ...int))
 	}
 
 	Interface struct {

--- a/internal/services/checker/deepscan/searcher_find_implementations.go
+++ b/internal/services/checker/deepscan/searcher_find_implementations.go
@@ -140,7 +140,7 @@ func (s *Searcher) applyMethodImplementationsInPackages(method *InjectionMethod,
 					}
 
 					for idx := gate.Index; idx <= maxIdx; idx++ {
-						callParam := callExpr.Args[gate.Index]
+						callParam := callExpr.Args[idx]
 
 						paramType := astPackage.TypesInfo.TypeOf(callParam)
 						targetName, targetPos, valid := s.extractTargetFromCallParam(paramType)

--- a/internal/services/checker/deepscan/searcher_find_implementations.go
+++ b/internal/services/checker/deepscan/searcher_find_implementations.go
@@ -131,43 +131,53 @@ func (s *Searcher) applyMethodImplementationsInPackages(method *InjectionMethod,
 				for gateIndex := range method.Gates {
 					gate := &method.Gates[gateIndex]
 					callMethod := callExpr.Fun
-					callParam := callExpr.Args[gate.Index]
-
-					paramType := astPackage.TypesInfo.TypeOf(callParam)
-					targetName, targetPos, valid := s.extractTargetFromCallParam(paramType)
-					if !valid {
-						// unknown injection type, possible generics or other not known
-						// go features on current moment
-						// we can extend this function later for new cases
+					if gate.IsVariadic && len(callExpr.Args) <= gate.Index {
 						continue
 					}
-
-					targetDefinitions := s.sourceFromToken(targetPos)
-					if targetDefinitions.Import == method.Definition.Import {
-						// injector use our public interface for typing
-						// we exclude this cases from more deep analyse, because of runtime
-						// types injection. (we have not known type on compile time)
-						continue
+					maxIdx := gate.Index
+					if gate.IsVariadic {
+						maxIdx = len(callExpr.Args) - 1
 					}
 
-					if !targetDefinitions.Place.Valid {
-						// invalid target
-						// possible is some not importable std const like `errors`
-						// or not known ast at this moment
-						continue
-					}
+					for idx := gate.Index; idx <= maxIdx; idx++ {
+						callParam := callExpr.Args[gate.Index]
 
-					gate.Implementations = append(gate.Implementations, Implementation{
-						Injector: Injector{
-							CodeName:         s.extractCodeFromASTNode(callParam),
-							ParamDefinition:  s.sourceFromToken(callParam.Pos()),
-							MethodDefinition: s.sourceFromToken(callMethod.Pos()),
-						},
-						Target: Target{
-							StructName: targetName,
-							Definition: targetDefinitions,
-						},
-					})
+						paramType := astPackage.TypesInfo.TypeOf(callParam)
+						targetName, targetPos, valid := s.extractTargetFromCallParam(paramType)
+						if !valid {
+							// unknown injection type, possible generics or other not known
+							// go features on current moment
+							// we can extend this function later for new cases
+							continue
+						}
+
+						targetDefinitions := s.sourceFromToken(targetPos)
+						if targetDefinitions.Import == method.Definition.Import {
+							// injector use our public interface for typing
+							// we exclude this cases from more deep analyse, because of runtime
+							// types injection. (we have not known type on compile time)
+							continue
+						}
+
+						if !targetDefinitions.Place.Valid {
+							// invalid target
+							// possible is some not importable std const like `errors`
+							// or not known ast at this moment
+							continue
+						}
+
+						gate.Implementations = append(gate.Implementations, Implementation{
+							Injector: Injector{
+								CodeName:         s.extractCodeFromASTNode(callParam),
+								ParamDefinition:  s.sourceFromToken(callParam.Pos()),
+								MethodDefinition: s.sourceFromToken(callMethod.Pos()),
+							},
+							Target: Target{
+								StructName: targetName,
+								Definition: targetDefinitions,
+							},
+						})
+					}
 				}
 			})
 		}

--- a/internal/services/checker/deepscan/searcher_find_implementations.go
+++ b/internal/services/checker/deepscan/searcher_find_implementations.go
@@ -131,11 +131,15 @@ func (s *Searcher) applyMethodImplementationsInPackages(method *InjectionMethod,
 				for gateIndex := range method.Gates {
 					gate := &method.Gates[gateIndex]
 					callMethod := callExpr.Fun
+
+					// variadic function arguments may be absent in callExpr.Args
 					if gate.IsVariadic && len(callExpr.Args) <= gate.Index {
 						continue
 					}
 					maxIdx := gate.Index
 					if gate.IsVariadic {
+						// variadic function arguments are passed as individual elements in callExpr.Args
+						// iterate over callExpr.Args to process all variadic arguments
 						maxIdx = len(callExpr.Args) - 1
 					}
 

--- a/internal/services/checker/deepscan/searcher_parse_methods.go
+++ b/internal/services/checker/deepscan/searcher_parse_methods.go
@@ -84,6 +84,11 @@ func (s *Searcher) extractMethodGates(astPackage *packages.Package, method *ast.
 				pos = field.Pos()
 			}
 
+			var isVariadic bool
+			if _, ok := field.Type.(*ast.Ellipsis); ok {
+				isVariadic = true
+			}
+
 			params = append(params, Gate{
 				MethodName:         method.Name.Name,
 				ParamName:          fieldIdent.Name,
@@ -95,6 +100,7 @@ func (s *Searcher) extractMethodGates(astPackage *packages.Package, method *ast.
 					Definition: s.sourceFromToken(pos),
 					GoType:     paramType.String(),
 				},
+				IsVariadic: isVariadic,
 			})
 		}
 	}

--- a/test/check/arch3_variadic.ct
+++ b/test/check/arch3_variadic.ct
@@ -1,0 +1,8 @@
+$ go-arch-lint check --project-path ${PWD}/test/check/project --arch-file arch3_variadic.yml --output-color=false
+module: github.com/fe3dback/go-arch-lint/test/check/project
+linters:
+   On | Base: component imports # always on
+   On | Advanced: vendor imports # switch 'allow.depOnAnyVendor = false' (or delete) to on
+   On | Advanced: method calls and dependency injections # switch 'allow.deepScan = true' (or delete) to on
+
+OK - No warnings found

--- a/test/check/project/arch1_nested_glob.yml
+++ b/test/check/project/arch1_nested_glob.yml
@@ -6,6 +6,7 @@ allow:
 exclude:
   - internal/excluded
   - vendor
+  - variadic
 
 excludeFiles:
   - "^.*_test\\.go$"

--- a/test/check/project/arch1_ok.yml
+++ b/test/check/project/arch1_ok.yml
@@ -6,6 +6,7 @@ allow:
 exclude:
   - internal/excluded
   - vendor
+  - variadic
 
 excludeFiles:
   - "^.*_test\\.go$"

--- a/test/check/project/arch1_warnings.yml
+++ b/test/check/project/arch1_warnings.yml
@@ -6,6 +6,7 @@ allow:
 exclude:
   - internal/excluded
   - vendor
+  - variadic
 
 excludeFiles:
   - "^.*_test\\.go$"

--- a/test/check/project/arch2_ok_fallback.yml
+++ b/test/check/project/arch2_ok_fallback.yml
@@ -6,6 +6,7 @@ allow:
 exclude:
   - internal/excluded
   - vendor
+  - variadic
 
 excludeFiles:
   - "^.*_test\\.go$"

--- a/test/check/project/arch3_variadic.yml
+++ b/test/check/project/arch3_variadic.yml
@@ -1,0 +1,19 @@
+version: 3
+
+workdir:
+  variadic
+
+allow:
+  depOnAnyVendor: false
+  deepScan: true
+
+components:
+  foo: { in: foo }
+  use_foo: { in: use_foo }
+
+deps:
+  foo:
+    anyProjectDeps: true
+
+  use_foo:
+    anyProjectDeps: true

--- a/test/check/project/variadic/foo/error.go
+++ b/test/check/project/variadic/foo/error.go
@@ -1,0 +1,11 @@
+package foo
+
+import (
+	"fmt"
+)
+
+type Foo []string
+
+func FooF(message string, args ...any) Foo {
+	return Foo{fmt.Sprintf(message, args)}
+}

--- a/test/check/project/variadic/use_foo/server.go
+++ b/test/check/project/variadic/use_foo/server.go
@@ -1,0 +1,11 @@
+package use_foo
+
+import (
+	"github.com/fe3dback/go-arch-lint/test/check/project/variadic/foo"
+)
+
+var (
+	foo1 = foo.FooF("boo")
+	foo2 = foo.FooF("boo", 5)
+	foo3 = foo.FooF("boo", 5, "test", 15)
+)

